xradar/io/backends/nexrad_level2.py

```diff
@@ -0,0 +1,1381 @@
+    ("reserved", INT4),
+    ("ngates", INT2),
+    ("first_gate", SINT2),
+    ("gate_spacing", SINT2),
+    ("thresh", SINT2),
+    ("snr_thres", SINT2),
+    ("flags", CODE1),
+    ("word_size", INT1),
+    ("scale", REAL4),
+    ("offset", REAL4),
+    # then data
+)
+
+# Table XVII-E Data Block (Volume Data Constant Type)
+# page 3-92
+VOLUME_DATA_BLOCK = (
+    ("block_type", "1s"),
+    ("data_name", "3s"),
+    ("lrtup", INT2),
+    ("version_major", INT1),
+    ("version_minor", INT1),
+    ("lat", REAL4),
+    ("lon", REAL4),
+    ("height", SINT2),
+    ("feedhorn_height", INT2),
+    ("refl_calib", REAL4),
+    ("power_h", REAL4),
+    ("power_v", REAL4),
+    ("diff_refl_calib", REAL4),
+    ("init_phase", REAL4),
+    ("vcp", INT2),
+    ("spare", "2s"),
+)
+
+# Table XVII-F Data Block (Elevation Data Constant Type)
+# page 3-93
+ELEVATION_DATA_BLOCK = (
+    ("block_type", "1s"),
+    ("data_name", "3s"),
+    ("lrtup", INT2),
+    ("atmos", SINT2),
+    ("refl_calib", REAL4),
+)
+
+# Table XVII-H Data Block (Radial Data Constant Type)
+# pages 3-93
+RADIAL_DATA_BLOCK = (
+    ("block_type", "1s"),
+    ("data_name", "3s"),
+    ("lrtup", INT2),
+    ("unambig_range", SINT2),
+    ("noise_h", REAL4),
+    ("noise_v", REAL4),
+    ("nyquist_vel", SINT2),
+    ("spare", "2s"),
+)
+
+
+def _find_range_params(scan_info):
+    """Return range parameters, first_gate, gate_spacing, last_gate."""
+    min_first_gate = 999999
+    min_gate_spacing = 999999
+    max_last_gate = 0
+    for scan_params in scan_info:
+        ngates = scan_params["ngates"][0]
+        for i, moment in enumerate(scan_params["moments"]):
+            first_gate = scan_params["first_gate"][i]
+            gate_spacing = scan_params["gate_spacing"][i]
+            last_gate = first_gate + gate_spacing * (ngates - 0.5)
+
+            min_first_gate = min(min_first_gate, first_gate)
+            min_gate_spacing = min(min_gate_spacing, gate_spacing)
+            max_last_gate = max(max_last_gate, last_gate)
+    return min_first_gate, min_gate_spacing, max_last_gate
+
+
+def format_nexrad_level2_data(
+    filename,
+    delay_field_loading=False,
+    station=None,
+    scans=None,
+    storage_options={"anon": True},
+    first_dimension=None,
+    group=None,
+    **kwargs,
+):
+    # Load the data file in using NEXRADLevel2File Class
+    nfile = NEXRADLevel2File(
+        prepare_for_read(filename, storage_options=storage_options)
+    )
+
+    # Access the scan info and load in the available moments
+    scan_info = nfile.scan_info(scans)
+    available_moments = {m for scan in scan_info for m in scan["moments"]}
+
+    # Deal with time
+    time_start, _time = nfile.get_times(scans)
+    time = get_time_attrs(time_start)
+    time["data"] = _time
+
+    # range
+    _range = get_range_attrs()
+    first_gate, gate_spacing, last_gate = _find_range_params(scan_info)
+    _range["data"] = np.arange(first_gate, last_gate, gate_spacing, "float32")
```

Comment:
Oh my, this is really hard to move from the CfRadial1 to CfRadial2 data model. AFAICS `_range`-dict is used for all sweeps (assuming range interpolated to a common grid). This would need to be done on a per sweep basis.

This would need another round of refactoring here.